contracts/src/v0.8/dev/ocr2dr/OCR2DRRegistry.sol

```diff
@@ -0,0 +1,783 @@
+  AggregatorV3Interface public immutable LINK_ETH_FEED;
+
+  // We need to maintain a list of consuming addresses.
+  // This bound ensures we are able to loop over them as needed.
+  // Should a user require more consumers, they can use multiple subscriptions.
+  uint16 public constant MAX_CONSUMERS = 100;
+  // 5k is plenty for an EXTCODESIZE call (2600) + warm CALL (100)
+  // and some arithmetic operations.
+  uint256 private constant GAS_FOR_CALL_EXACT_CHECK = 5_000;
+  // Maximum number of oracles DON can support
+  // Needs to match OCR2Abstract.sol
+  uint256 internal constant MAX_NUM_ORACLES = 31;
+
+  error TooManyConsumers();
+  error InsufficientBalance();
+  error InvalidConsumer(uint64 subId, address consumer);
+  error InvalidSubscription();
+  error OnlyCallableFromLink();
+  error InvalidCalldata();
+  error MustBeSubOwner(address owner);
+  error MustBeAllowedDon();
+  error PendingRequestExists();
+  error MustBeRequestedOwner(address proposedOwner);
+  error BalanceInvariantViolated(uint256 internalBalance, uint256 externalBalance); // Should never happen
+  event FundsRecovered(address to, uint256 amount);
+
+  struct Subscription {
+    // There are only 1e9*1e18 = 1e27 juels in existence, so the balance can fit in uint96 (2^96 ~ 7e28)
+    uint96 balance; // Common link balance used for all consumer requests.
+  }
+  // We use the config for the mgmt APIs
+  struct SubscriptionConfig {
+    address owner; // Owner can fund/withdraw/cancel the sub.
+    address requestedOwner; // For safely transferring sub ownership.
+    // Maintains the list of keys in s_consumers.
+    // We do this for 2 reasons:
+    // 1. To be able to clean up all keys from s_consumers when canceling a subscription.
+    // 2. To be able to return the list of all consumers in getSubscription.
+    // Note that we need the s_consumers map to be able to directly check if a
+    // consumer is valid without reading all the consumers from storage.
+    address[] consumers;
+  }
+  // Note a nonce of 0 indicates an the consumer is not assigned to that subscription.
+  mapping(address => mapping(uint64 => uint64)) /* consumer */ /* subId */ /* nonce */
+    private s_consumers;
+  mapping(uint64 => SubscriptionConfig) /* subId */ /* subscriptionConfig */
+    private s_subscriptionConfigs;
+  mapping(uint64 => Subscription) /* subId */ /* subscription */
+    private s_subscriptions;
+  // We make the sub count public so that its possible to
+  // get all the current subscriptions via getSubscription.
+  uint64 private s_currentSubId;
+  // s_totalBalance tracks the total link sent to/from
+  // this contract through onTokenTransfer, cancelSubscription and oracleWithdraw.
+  // A discrepancy with this contract's link balance indicates someone
+  // sent tokens using transfer and so we may need to use recoverFunds.
+  uint96 private s_totalBalance;
+  event SubscriptionCreated(uint64 indexed subId, address owner);
```

Comment:
I can see a little mess of `subId` and `subscriptionId`, I would suggest we always use `subscriptionId` for clarity.